jam/server/api/v2/base.py

```diff
@@ -0,0 +1,296 @@
+            err = None
+
+        self._view = self._view_class(*loaded)
+
+        if 'relationship' in self.path_kwargs:
+            relationship = self._serializer.relations[self.path_kwargs['relationship']]
+            self._view = relationship.view(*loaded)
+            self._serializer = relationship.serializer()
+
+        permissions = Permissions.get_permissions(self.current_user, *loaded)
+        required_permissions = self._view.get_permissions(self.request)
+
+        # For use later on
+        self.current_user.permissions = permissions
+
+        if (required_permissions & permissions) != required_permissions:
+            if self.current_user.uid is None:
+                raise exceptions.Unauthorized()
+            raise exceptions.Forbidden(required_permissions)
+
+        if err:
+            raise err
+
+        if self.request.method in ('GET', 'DELETE'):
+            return  # GET and DELETE bodies are ignored
+
+        if not isinstance(self.json, dict):
+            raise exceptions.MalformedData()
+
+        data = self.json.get('data', {})
+        if data.get('type') != self._view.plural:
+            raise exceptions.IncorrectParameter('data.type', self._view.type, data.get('type', 'null'))
+
+    # Create
+    def post(self, **args):
+        if self._view.resource:
+            raise tornado.web.HTTPError(http.client.METHOD_NOT_ALLOWED)
+
+        if isinstance(self.json['data'], list):
+            return self.post_bulk()
+
+        new = self._view.create(self.json['data'].get('id'), self.json['data']['attributes'], self.current_user)
+        self.write({'data': self.serialize(new)})
+        self.set_status(http.client.CREATED)
+
+    def post_bulk(self):
+        if 'bulk' not in self.extensions:
+            raise Exception()
+        new, errors = [], []
+        for entry in self.json['data']:
+            try:
+                new.append(self._view.create(entry.get('id'), entry['attributes'], self.current_user))
+            except exceptions.JamException as e:
+                new.append(None)
+                error.append(e.serialize())
+            else:
+                error.append(None)
+        self.write({
+            'data': [self.serialize(n) for n in new],
+            'errors': errors
+        })
+        self.set_status(http.client.CREATED)
+
+    # Read/List
+    def get(self, **args):
+        if self._view.resource:
+            return self.write({
+                'data': self.serialize(self._view.read(self.current_user))
+            })
+
+        selector = self._view.list(self.filter, self.sort, self.page, self.page_size, self.current_user)
+
+        return self.write({
+            'meta': {
+                'total': selector.count(),
+                'perPage': self.page_size
+            },
+            # TODO
+            'links': {},
+            'data': [self.serialize(resource) for resource in selector]
+        })
+
+    def put(self, **args):
+        return super().put()
```

Comment:
Super just swallows the args and raises a 405. This effectively a todo as PUT is in a grey area for JSONAPI.